hoot-core/src/main/cpp/hoot/core/io/OsmApiChangeset.cpp

```diff
@@ -1657,34 +1658,36 @@ void XmlChangeset::failRelation(long id, bool beforeSend)
   LOG_TRACE("Failed relation (" << id << ")");
 }
 
-void XmlChangeset::getNodes(const ChangesetInfoPtr& changeset, QTextStream& ts, ChangesetType type, long changeset_id)
+void XmlChangeset::writeElements(const ChangesetInfoPtr& changeset, QTextStream& ts, ChangesetType type, long changeset_id,
+                                 ElementType::Type elementType, const ChangesetElementMap& elements)
 {
-  XmlElementMap& nodes = _nodes[type];
-  for (ElementIdToIdMap::iterator it = _idMap.begin(ElementType::Node); it != _idMap.end(ElementType::Node); ++it)
+  vector<long> outputElements;
+  //  Iterate only those elements that are in the small changeset to be sent and not all of the elements in entire changeset
+  for (ChangesetInfo::iterator it = changeset->begin(elementType, type); it != changeset->end(elementType, type); ++it)
   {
-    if (nodes.find(it->second) != nodes.end() && changeset->contains(ElementType::Node, type, it->second))
-      ts << nodes.at(it->second)->toString(changeset_id);
+    if (elements.find(*it) != elements.end() && _idMap.containsId(elementType, *it))
+      outputElements.push_back(_idMap.getId(elementType, *it));
   }
+  //  Sort the elements by ID
+  std::sort(outputElements.begin(), outputElements.end(), id_sort_order);
+  //  Iterate the elements again, writing them to the output stream
+  for (vector<long>::iterator it = outputElements.begin(); it != outputElements.end(); ++it)
+    ts << elements.at(_idMap.getId(elementType, *it))->toString(changeset_id);
 }
```

Comment:
This is where one of the big changes occurs.  Before we iterated all nodes/ways/relations in the entire larger changeset looking for the ones that exist in the smaller sub-changeset that we are about to send.  Now we iterate only the nodes/ways/relations in the sub-changeset (which is a smaller number of iterations), then we sort them, then finally output them.  It requires iterating them twice but `2n << m` where n is the number in the smaller changeset, 10k at most, and m is the larger changeset, potentially millions.